Build and deployment documentation

ccarouge opened the pull request:
<!-- This template was created to cover the most complex case of new changes to the configuration. As such, all sections may not be suitable for other types of changes (e.g. documentation changes, cherry-picking changes between configurations). Please fill in all the sections you believe are suitable to your case. The reviewer(s) will ask for any missing information. Please do not delete any empty sections. -->

**1. Summary**:

Added documentation about the building and deployment process for ACCESS-ESM1.6

**2. Issues Addressed:**
<!-- Add links to github issue(s) this is related to -->
- Fixes #279 

**3. Dependencies (e.g. on payu, or model)**

This change requires changes to (note pull request(s) where relevant):
- [ ] workflow manager (payu):
- [ ] model deployment (ACCESS-ESM1.6):
- [ ] model component or library dependency:
- [ ] input workflow:

<!-- Describe and link to the related changes to dependencies -->

**4. Ad-hoc Testing**

What ad-hoc testing was done? How are you convinced this change is correct (plots are good)?

Documentation only. No testing done. Relying on build preview in PR to correct formatting.

**5. CI Testing**
<!-- Has the CI-testing been run? -->
- [ ] `!test repro` has been run
- [X] N/A: documentation only

**6. Reproducibility**

Is this reproducible with the previous commit? (If not, why not?)
- [X] Yes
- [ ] No - `!test repro commit` has been run. <!-- add detail below for why it's answer changing -->

**8. Manifests**

Have you changed the executable, the input files and/or the restart files?
- [ ] Yes
- [X] No

If yes, have you updated the manifests?
- [ ] Yes
- [ ] No

To update the manifests, run payu setup (in a cloned copy of your feature branch) with reproducibility tests turned off:

```yaml
manifest:
  reproduce:
    exe: false
    input: false
    restart: false
runlog:
  enable: false
```
Then commit the newly created manifest files (under manifests/) only to the branch for this PR.


**7. Documentation**
<!--Does this impact documentation? -->

Is the documentation updated?
- [X] Yes
- [ ] N/A

**9. Merge Strategy**
<!-- What is the planned merge strategy (Merge commit, Rebase and merge, or squash) ?
If not squash, link to the related issue in the commit descriptions -->

- [ ] Merge commit
- [ ] Rebase and merge
- [X] Squash

ccarouge committed:
Build and deployment first draft

atteggiani commented:
> @atteggiani The config docs for ESM1.6 fail to deploy with the preview. Does it need all the PR previews to work in order to deploy? I'm happy to go and fix PR 252, just making sure that should solve the problem before investing time.

Yes it needs all PR previews to work, unfortunately. 
If the fix takes time, you might try writing a line with `!no-pr-preview` in the body of the PR (PR #252 in this case). It should not trigger that PR preview. I haven't tested this properly due to RTD changes so it might not work, but it's worth a try. 
It needs to be the first line of the PR body.

Then you would need to re-trigger this PR. You cannot simply retrigger the build manually, so you'll have to add a new commit or amend (no edits) and force push the latest one.

chrisb13 commented:
Thanks @ccarouge, fwiw, you might find some of this material helpful:
 - https://access-om3-configs.access-hive.org.au/infrastructure/Building/
 - https://access-community-hub.github.io/access-iom3-configs/infrastructure/Building/

ccarouge commented:
PR #252 was fixed and the preview workflow ran from #252  has now created a preview (dismiss the failed build in this PR).

@chrisb13 I'll have a look at the links you shared.

ccarouge committed:
Minor updates

Added information on required repositories and building methods for ACCESS-ESM1.6.

ccarouge committed:
(#279) - Clarify the prerequisites and the recommended build workflow

ccarouge committed:
(#279) - Revisiting the page completely.

ccarouge committed:
(#279) - updates after reading the rendered version.

ccarouge commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+
+    The _UM7_ and _GCOM4_ repositories are private because these software are not open-source. They are covered by the UKMO's Momentum Partnership licence. To gain access to these repositories, please request an invitation via [this ACCESS-Hive forum topic](https://forum.access-hive.org.au/t/request-access-to-access-esm1-6-component-repositories/5709).
+
+## Building ACCESS-ESM1.6 using the deployment infrastructure
```
@aidanheerdegen @harshula , I'd appreciate if one of you could have a read through this section before the end of January. I want to make sure I don't misconstrue anything.

ccarouge commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -1,11 +1,23 @@
-If you need to modify the source code for ACCESS-ESM1.6, you will need to use the Spack infrastructure to create your own build. There are two ways to do so (on GitHub or locally) both with benefits and limitations.
+In order to build your own version of ACCESS-ESM1.6, you will need read access to the following repositories: 
+
+- https://github.com/ACCESS-NRI/cice5
```
Done. But I have completely redesigned the build page.

ccarouge committed:
(#279) - Add information on existing deployments

ccarouge committed:
(#279) - Reorganise info between Building and Deployment pages.

aidanheerdegen reviewed (dismissed):
A couple of comments, but otherwise seems good, thanks.

ccarouge committed:
(#279) - Integrated Aidan's feedback

ccarouge committed:
(#279) - add deployment page to navigation and move navigation entry higher up in mkdocs.yml

Got tired of having to scroll down or search for the nav entry when it's the bit we modify the most.

ccarouge committed:
(#279) - Re-read

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+
+## Software structure
+
+ACCESS-ESM1.6 being a Earth System model, it is constituted of several model components, each residing in its own repository on GitHub. In addition, some model components need other utilities at compilation time that are stored in separate repositories. This results in a complex software structure involving a large number of repositories:
```
```suggestion
ACCESS-ESM1.6 being a Earth System model, it is constituted of several model components, each residing in its own repository on GitHub. In addition, some model components need other utilities at compilation time that are stored in separate repositories. This results in a complex software structure involving a number of repositories, including:
```

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+# Build infrastructure
+
+Most users of ACCESS-ESM1.6 will not need to build the model. Released versions of the model are deployed by ACCESS-NRI to Gadi and available to users. Check the [Deployment page](/documentation/docs/pages/infrastructure/Deployment.md) to learn about available deployments. If you require your own build of the model, typically because you need to modify one or several model components, you will fi ...
+
```
Do we need to define release and deploy, we use them both, slightly interchangeably

anton-seaice commented on documentation/mkdocs.yml:
```diff
@@ -10,7 +10,36 @@ site_url: !ENV READTHEDOCS_CANONICAL_URL
+        - Deployment:            pages/infrastructure/Deployment.md
+        - Building:              pages/infrastructure/Building.md
+        - Git practices:         pages/infrastructure/Git-practices.md
+
```
why did this section get moved ?

ccarouge replied:
It's the section we update the most, there is no reason for it to be so far down the file. I'm tired of searching or scrolling for it.

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+
+## Local build of ACCESS-ESM1.6
+
+The Spack infrastructure allows you to create your own local build on _Gadi_ for ACCESS-ESM1.6. Once setup, this might be faster than using the deployment infrastructure, especially when you want a lot of small incremental builds during development.
```
```suggestion
The Spack infrastructure allows you to create your own local build on _Gadi_ for ACCESS-ESM1.6. Once setup, this is faster than using the deployment infrastructure, especially when writing new code.
```

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+
+The Spack infrastructure allows you to create your own local build on _Gadi_ for ACCESS-ESM1.6. Once setup, this might be faster than using the deployment infrastructure, especially when you want a lot of small incremental builds during development.
+
+However, this method has several drawbacks. It requires users to install the Spack software which can make it harder for ACCESS-NRI to provide support. It requires users to have at least _read permissions_ on all the repositories required by ACCESS-ESM1.6 (see [Software structure](#software-structure)). It builds the model in the user space on _Gadi_ thus potentially exposing licensed source code ...
```
```suggestion
However, this method has several drawbacks. It requires users to install the Spack software which will take some investment of time to understand. It is also harder to share and collaborate on changes, including making it harder for ACCESS-NRI to provide support. It requires users to have at least _read permissions_ on all the repositories required by ACCESS-ESM1.6 (see [Software structure](#software-structure)). 
```

anton-seaice replied:
I don't think number of inodes are a real problem for an individual developer. Also if folks have access to the UM source code then they have the appropriate permissions to see the source code.

anton-seaice commented on documentation/docs/pages/infrastructure/Deployment.md:
```diff
@@ -0,0 +1,29 @@
+
+There are development releases (identified with a `dev_` prefix to the date tag) and full releases (identified with only a date tag). We recommend new users start with the newest full release (i.e., with the most recent date tag). 
+
+These builds are optimised for the Sapphire-Rapid hardware from Intel (i.e., the *normalsr* queue), and may not run on Cascade-Lake (i.e., the *normal* queue) or older hardware.
```
I'm pretty confident it would run fine on normal. I can't see any HW specific optimisation going on

anton-seaice commented on documentation/docs/pages/infrastructure/Deployment.md:
```diff
@@ -0,0 +1,29 @@
+# Deployment
+
+All released versions of ACCESS-ESM1.6 come with a pre-deployed build on _Gadi_. Deployed builds are stored under the project `vk83` (hence requiring [membership to this project](https://my.nci.org.au/mancini/project/vk83/join) for access). You can see all deployed builds via:
```
Mostly the words a "release version" would refer to a configuration, not a model exe

I think the point here is - release configurations use a certain deployment of the model code. Mostly users wouldn't mess with that. 

Up to date configs will use the latest deployment, the older ones are kept for folks who are running experiments based on older configurations

ccarouge replied:
I would say the CalVer on the model build is a released version. We don't advertise released builds directly and only advertise released configurations but it does not mean we don't release the builds. I'm happy to add configurations in the text, I just wanted to say for me we release both but only advertise one.

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,44 @@
+# Build infrastructure
+
+Most users of ACCESS-ESM1.6 will not need to build the model. Released versions of the model are deployed by ACCESS-NRI to Gadi and available to users. Check the [Deployment page](/documentation/docs/pages/infrastructure/Deployment.md) to learn about available deployments. If you require your own build of the model, typically because you need to modify one or several model components, you will fi ...
```
The link to Deployment page doesn't work , this might be correct
```suggestion
Most users of ACCESS-ESM1.6 will not need to build the model. Released versions of the model are deployed by ACCESS-NRI to Gadi and available to users. Check the [Deployment page](infrastructure/Deployment.md) to learn about available deployments. If you require your own build of the model, typically because you need to modify one or several model components, you will find the required information on this Builing page.
```

anton-seaice reviewed (commented):
Sorry a bit of a chaotic review, I can have another look as needed

ccarouge committed:
(#279) - Edits from review

ccarouge committed:
(#279) - More edits.

ccarouge commented:
Thanks for the review @anton-seaice . I've replied to a few comments and I have a modified version of the text for review.

anton-seaice commented on documentation/docs/pages/infrastructure/Building.md:
```diff
@@ -0,0 +1,52 @@
+
+!!! note
+
+    ACCESS-ESM1.6 has other dependencies such as Fortran compilers, netCDF and more. These are not listed above as they are more generic tools that we never modify. The build infrastructure we provide will find these dependencies for you.
```
```suggestion
    ACCESS-ESM1.6 has other dependencies such as Fortran compilers, netCDF and more. These dependencies are considered model infrastructure are rarely changed by end-users. The build infrastructure we provide will find these dependencies for you.
```

anton-seaice reviewed (commented):
Looks - good . Some minor refinements from me

ccarouge committed:
(#279) - Edits from review

ccarouge committed:
(#279) - links check

ccarouge committed:
(#279) - Link check

ccarouge committed:
(#279) - use lowercase for all documentation files

Consistency makes linking between pages easier.

ccarouge committed:
(#279) - Update links to lowercase filenames

anton-seaice reviewed (approved):
Thanks Claire !
